### PR TITLE
functional equivalence options

### PR DESCRIPTION
**Pull request by dozy**
seq_alignment changes for functional equivalence options
 - duplicate marking
   - set the markdup_method parameter (biobambam|samtools|picard, default:biobambam)
   - detect flowcell type (patterned or not) and set the pixel distance for optical duplicate detection

 - bwakit postalt processing:
     - can be enabled via the product_release.yml file or command-line flag (alt haplotype reference required)
     - path to postalt javascript script must be specified via the NPG_PIPELINE_JS_SCRIPTS_DIR environment variable or via general_values

 - fix and update tests

Note: depends on https://github.com/wtsi-npg/p4/pull/260 and https://github.com/wtsi-npg/p4/pull/259 (https://github.com/wtsi-npg/p4/pull/259 only needed if using bwakit postalt processing)

### CODE REVIEW

File: lib/npg_pipeline/function/seq_alignment.pm [hunk 1]
```diff
@@ -314,9 +337,19 @@ sub _alignment_command { ## no critic (Subroutines::ProhibitExcessComplexity)
...
+  # processing is enabled, then we will use bwa's "mem" with post-processing using
+  # the bwa-postalt.js script from bwakit
+  if($do_target_alignment and ($self->bwakit or $self->bwakit_enable($dp))) { # two ways to specify bwakit?
+    if((my $alt_ref = $self->_alt_reference($dp))) {
+      $p4_param_vals->{alignment_method} = $bwa = 'bwa_mem_bwakit';
+      $p4_param_vals->{fa_alt_path} = $alt_ref;
+      $p4_param_vals->{js_dir} = $self->_js_scripts_dir;
+    }
+    else {
+      $self->info(q[bwakit postal processing specified, but no alternate haplotypes in reference]);
```
Comment (mp15) [inline lib/npg_pipeline/function/seq_alignment.pm]: s/postal/postalt/g

Comment (dozy) [reply lib/npg_pipeline/function/seq_alignment.pm]: fixed

**Comment by mp15** [comment]
Any chance you might be willing to rebase this? It's becoming a bit cumbersome to apply the patches otherwise.

**Comment by ces** [comment]
bwakit untested.

**Pull request merged**